bump golangci-lint v1.20.0

thaJeztah opened the pull request:
release notes: 

- v1.20.0: https://github.com/golangci/golangci-lint/releases/tag/v1.20.0
    - Add WSL linter (#771)
    - Add gocognit linter (#756)
    - govet: add more analyzers
    - dramatically reduce memory usage
    - reduce 1.5x memory usage on large repos on repeated runs
    - Rename deadline option to timeout and mark deadline as deprecated
- v1.19.1: https://github.com/golangci/golangci-lint/releases/tag/v1.19.1
- v1.19.0: https://github.com/golangci/golangci-lint/releases/tag/v1.19.0
    - Add dogsled linter.
    - Add whitespace linter
    - Add Godox linter support
    - Various performance optimizations
- v1.18.0: https://github.com/golangci/golangci-lint/releases/tag/v1.18.0
    - Support go1.13
    - Update format of junit xml output to mark failures as such
    - Speed up linting: use deduplicated packages
    - Add user supplied error messages in depguard issues
    - Add funlen linter

thaJeztah committed:
bump golangci-lint v1.20.0

release notes:

- v1.20.0: https://github.com/golangci/golangci-lint/releases/tag/v1.20.0
    - Add WSL linter (#771)
    - Add gocognit linter (#756)
    - govet: add more analyzers
    - dramatically reduce memory usage
    - reduce 1.5x memory usage on large repos on repeated runs
    - Rename deadline option to timeout and mark deadline as deprecated
- v1.19.1: https://github.com/golangci/golangci-lint/releases/tag/v1.19.1
- v1.19.0: https://github.com/golangci/golangci-lint/releases/tag/v1.19.0
    - Add dogsled linter.
    - Add whitespace linter
    - Add Godox linter support
    - Various performance optimizations
- v1.18.0: https://github.com/golangci/golangci-lint/releases/tag/v1.18.0
    - Support go1.13
    - Update format of junit xml output to mark failures as such
    - Speed up linting: use deduplicated packages
    - Add user supplied error messages in depguard issues
    - Add funlen linter

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Looks like RS5 hung at some point; https://ci.docker.com/public/job/moby/job/PR-39964/3/execution/node/179/log/


```
01:06:11.788  === RUN   TestDockerSuite/TestRestartAutoRemoveContainer
02:00:03.559  Sending interrupt signal to process
02:00:12.237  Sending interrupt signal to process
02:00:21.976  exit status 4294967295
02:00:21.976  FAIL	github.com/docker/docker/integration-cli	5356.357s
02:00:23.559  After 20s process did not stop
```

thaJeztah commented:
Looks like it's more memory-hungry than the previous version;

```
[2019-09-23T13:51:57.344Z] fatal error: runtime: out of memory
[2019-09-23T13:51:57.344Z] 
[2019-09-23T13:51:57.344Z] runtime stack:
[2019-09-23T13:51:57.344Z] runtime.throw(0x56332a6d970f, 0x16)
[2019-09-23T13:51:57.344Z] 	/usr/local/go/src/runtime/panic.go:774 +0x74
[2019-09-23T13:51:57.344Z] runtime.sysMap(0xc178000000, 0x4000000, 0x56332b5a3b58)
[2019-09-23T13:51:57.344Z] 	/usr/local/go/src/runtime/mem_linux.go:169 +0xc7
[2019-09-23T13:51:57.344Z] runtime.(*mheap).sysAlloc(0x56332b5815e0, 0x2000, 0x563329e1d444, 0x7f6939f4a008)
[2019-09-23T13:51:57.344Z] 	/usr/local/go/src/runtime/malloc.go:701 +0x1cf
[2019-09-23T13:51:57.344Z] runtime.(*mheap).grow(0x56332b5815e0, 0x1, 0xffffffff)
[2019-09-23T13:51:57.344Z] 	/usr/local/go/src/runtime/mheap.go:1252 +0x44
```

kolyshkin commented:
> Looks like it's more memory-hungry than the previous version

and 1.19 is even worse than that, at least according to https://github.com/golangci/golangci-lint/issues/731

I wish there would be a way to somehow restrict its memory consumption (I think gometalinter had something)

kolyshkin commented:
I also tried 1.19.1 and it also uses more memory (5GB or more) than the currently used 1.17.1 (4 GB).

So I guess we have to wait a bit more (and participate in https://github.com/golangci/golangci-lint/issues/337).

thaJeztah commented:
the ticket mentioned a new linter, so perhaps we should skip that one for now, and don't use the default set of linters as a starting point (as it may change).

We should add a note though to review the list of linters when updating, otherwise we may miss out on new linters that could be good to have.

Let me also check if we can work around this by reducing the number of linters run in parallel

kolyshkin commented:
> Let me also check if we can work around this by reducing the number of linters run in parallel

I tried removing govet entirely (as per recommendation in https://github.com/golangci/golangci-lint/issues/337#issuecomment-535024279) but it did not made any difference at all.

kolyshkin commented:
Adding `disable-all: true` to `linters:` section indeed removes a few linters (staticcheck, structcheck, typecheck, unused, varcheck) but does not change much in terms of RAM used.

kolyshkin commented:
The default set (compared to what we have right now) excludes the following linters: `gofmt goimports golint gosec misspell unconvert`. Not much difference in memory usage either.

I wonder why we're using both gofmt and goimports? Afaik goimports is a superset of gofmt.

crosbymichael commented:
We had the memory issues in containerd and made the following changes

https://github.com/containerd/containerd/commit/57b51b94815aa2e592264eb69c2f98e3a6616adc

Increased the deadline to 2m and set GOGC=75 to help with the memory issues.

thaJeztah committed:
builder/dockerfile/mockbackend_test.go: suppress SA9005 (staticcheck)

```
builder/dockerfile/mockbackend_test.go:107:21: SA9005: struct doesn't have any exported fields, nor custom marshaling (staticcheck)
	return json.Marshal(rawImage(*i))
	                   ^
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
graphdriver/aufs: SA4021: x = append(y) is equivalent to x = y (staticcheck)

```
daemon/graphdriver/aufs/aufs_test.go:746:8: SA4021: x = append(y) is equivalent to x = y (staticcheck)
	ids = append(ids[2:])
	      ^
```

Also pre-allocating the ids slice while we're at it.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
distribution: SA4021: x = append(y) is equivalent to x = y (staticcheck)

```
distribution/push_v2_test.go:552:29: SA4021: x = append(y) is equivalent to x = y (staticcheck)
	return nil, errcode.Errors(append([]error{errcode.ErrorCodeUnauthorized.WithMessage("unauthorized")}))
	                           ^
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
fix nolint comments for SA1019: filters.ToParamWithVersion is deprecated

The old nolint comment didn't seem to work anymore;

```
client/container_list.go:39:22: SA1019: filters.ToParamWithVersion is deprecated: do not use in any new code; use ToJSON instead  (staticcheck)
client/events.go:94:22:         SA1019: filters.ToParamWithVersion is deprecated: do not use in any new code; use ToJSON instead  (staticcheck)
client/image_list.go:28:22:     SA1019: filters.ToParamWithVersion is deprecated: do not use in any new code; use ToJSON instead  (staticcheck)
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon/events/testutils: remove redundant variable

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/jsonmessage: fix SA1006: printf-style function with no arguments

Also fixed some incorrectly formatted comments

```
pkg/jsonmessage/jsonmessage.go:180:20: SA1006: printf-style function with dynamic format string and no further arguments should use print-style function instead (staticcheck)
		fmt.Fprintf(out, endl)
		                 ^
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
graphdriver/btrfs: SA4003: no value of type uint64 is less than 0 (staticcheck)

```
daemon/graphdriver/btrfs/btrfs.go:609:5: SA4003: no value of type uint64 is less than 0 (staticcheck)
	if driver.options.size <= 0 {
	   ^
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon: S1033: unnecessary guard around call to delete (gosimple)

```
daemon/container_operations.go:787:2: S1033: unnecessary guard around call to delete (gosimple)
	if _, ok := container.NetworkSettings.Networks[n.ID()]; ok {
	^
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
aufs: SA4011: did you mean to break out of the outer loop? (staticcheck)

As caught by staticcheck (after disabling the default exclusion rules);

Based on the comment, this break was indeed meant to break the
loop and return the error.

```
daemon/graphdriver/aufs/mount.go:54:4: SA4011: ineffective break statement. Did you mean to break out of the outer loop? (staticcheck)
			break
			^
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
pkg/mount: SA4011: ineffective break statement (staticcheck)

```
pkg/mount/mountinfo_linux.go:93:5: SA4011: ineffective break statement. Did you mean to break out of the outer loop? (staticcheck)
				break
				^
```

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
golangci-lint: update exclusion rules for todo's

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
golangci-lint: fix "golint" not ignoring generated files

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
golangci-lint: disable default exclusion rules

The default exclusion rules were too permissive; disable
them, but copy the relevant ones to the configuation.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
golangci-lint: include all enabled linters in config

This makes it clearer from the configuration what linters
are enabled.

Also disables the `gofmt` linter, because it's superseded
by the `goimports` linter.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Updated to v1.20.0, which fixes the memory issue, and improves performance.

After updating to v1.20.0, I tried some combinations;

```
GOGC=20, concurrency: 2

INFO File cache stats: 1459 entries of total size 6.9MiB 
INFO Memory: 295 samples, avg is 1761.0MB, max is 2425.3MB 
INFO Execution took 48.864976632s                 

GOGC=20, concurrency: 8 (default)

INFO File cache stats: 1460 entries of total size 6.9MiB 
INFO Memory: 378 samples, avg is 358.2MB, max is 2292.1MB 
INFO Execution took 48.611063372s                 

GOGC=20, concurrency: 10

INFO File cache stats: 1459 entries of total size 6.9MiB 
INFO Memory: 320 samples, avg is 1709.6MB, max is 2426.0MB 
INFO Execution took 54.095506023s                 
```


```
GOGC=75, concurrency: 2

INFO File cache stats: 1459 entries of total size 6.9MiB 
INFO Memory: 229 samples, avg is 2184.7MB, max is 3029.3MB 
INFO Execution took 35.750402921s 
```

```
GOGC=75, concurrency: 8 (default)

INFO File cache stats: 1460 entries of total size 6.9MiB 
INFO Memory: 205 samples, avg is 665.3MB, max is 2561.8MB 
INFO Execution took 28.021308374s                 
```

```
GOGC=75, concurrency: 10

INFO File cache stats: 1460 entries of total size 6.9MiB 
INFO Memory: 228 samples, avg is 613.0MB, max is 2696.2MB 
INFO Execution took 30.581860965s                 
```

Default

```
concurrency: 8 default (no GOGC or concurrency set)

INFO File cache stats: 1459 entries of total size 6.9MiB 
INFO Memory: 231 samples, avg is 2695.7MB, max is 3628.3MB 
INFO Execution took 33.262956944s                 
```



Based on the above, I think `GOGC=75, concurrency: 8 (default)` looks ok

kolyshkin commented on daemon/config/config.go:
```diff
@@ -493,9 +493,7 @@ func findConfigurationConflicts(config map[string]interface{}, flags *pflag.Flag
-				if _, valid := unknownKeys[namedOption.Name()]; valid {
-					delete(unknownKeys, namedOption.Name())
-				}
+				delete(unknownKeys, namedOption.Name())
```
(sorry, there's no way to comment the commit message itself, so putting this here)

The commit message from this commit mentions issue that is not here (SA4003), please remove

thaJeztah replied:
Good catch; copied a couple of lines too many. Updated 👍

kolyshkin commented on hack/dockerfile/install/golangci_lint.installer:
```diff
@@ -1,11 +1,19 @@
 install_golangci_lint() {
 	echo "Installing golangci-lint version ${GOLANGCI_LINT_COMMIT}"
 	go get -d github.com/golangci/golangci-lint/cmd/golangci-lint
 	cd "$GOPATH/src/github.com/golangci/golangci-lint/" || exit 1
```
side note; maybe we should do `set -u -e` at the beginning of this script, instead of using `|| exit 1` here.

thaJeztah replied:
I think we can in-line most of these scripts in the Dockerfile eventually

kolyshkin reviewed (approved):
LGTM except for a nit about a commit message

tiborvass commented on pkg/mount/mountinfo_linux.go:
```diff
@@ -90,7 +90,6 @@ func parseInfoFile(r io.Reader, filter FilterFunc) ([]*Info, error) {
 				   mount propagation flags in fields[6]. The correct
 				   behavior is to ignore any unknown optional fields.
 				*/
-				break
```
I was wondering whether this `break` meant break from the switch or break from the for loop, but i realize i don't understand the logic in this for loop at all:

If `i` starts at 6 and it's smaller than `numFields` and `fields[i] != "-"`  only then set p.Optional, otherwise increment i, which means it will never be equal to 6 again, so p.Optional will never be set. So this entire for loop could be reduced to:
`if numFields >= 6 && fields[6] != "-" { p.Optional = fields[6] }`

Surely I must be missing something here.

thaJeztah replied:
I think it's to increment `i` because that's used further down to validate if the output was correct, and to concatenation the remaining fields
